### PR TITLE
New: no-magic-number rule (fixes #4027)

### LINKED ISSUE #4027: New: rule no-magic-number
**When does this rule warn? Please describe and show example code:**

This rule will warn when magic numbers are used in the code. For example :
```javascript
// Valid code
var TAX_PERCENTAGE = 0.25;

var dutyFreePrice = 100,
    finalPrice = dutyFreePrice + (dutyFreePrice * TAX_PERCENTAGE);
```

```javascript
// Invalid code : Warns "No magic number: 0.25"
var dutyFreePrice = 100,
    finalPrice = dutyFreePrice + (dutyFreePrice * 0.25); 
```

The configuration takes 3 parameters :

* `ignore`: An array of numbers to ignore. It's set to `[0, 1, 2]` by default.
* `detectObjects`: A boolean to specify if we should detect numbers when setting object properties for example. `false` by default.
* `enforceConst`: A boolean to specify if we should check for the const keyword in variable declaration of numbers. `false` by default.

**Is this rule preventing an error or is it stylistic?**

It's more of a best practice / stylistic rule. Avoiding 'magic numbers' is good for understanding the code and refactoring it.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

I think it's a good candidate for inclusion because its application is really broad. "Magic numbers" and their usage is not something specific to my team, but concerns every developers (it even has its own [wikipedia article](https://en.wikipedia.org/wiki/Magic_number_(programming)) :) ).

**Are you willing to create the rule yourself?**

Yes :). Or to be more precise, I used @danielstjules's library, [buddy.js](https://github.com/danielstjules/buddy.js/), as the basis for this rule (code and specifications). I liked what buddy.js did, but I didn't want to use yet another tool for checking for magic numbers.
I will make it a plugin if your think it's not worth integrating it in core of course.

Cheers

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27139073-new-rule-no-magic-number?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by kombucha**
Proposed implementation of rule no-magic-number.
Fixes issue #4027

### CODE REVIEW

File: docs/rules/no-magic-number.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# Disallow Magic Numbers (no-magic-number)
+
+'Magic numbers' are numbers that occur multiple time in code without an explicit meaning.
+They should preferably be replaced by named constants.
+
+```js
+if (foo == null) {
+  bar();
+}
+```
```
Comment (btmills) [inline docs/rules/no-magic-number.md]: How is this example related to the rule?

Comment (kombucha) [inline docs/rules/no-magic-number.md]: It's not. I used another doc as a template and failed to clean it up properly :/. 
Will it make travis' build fail again if I fix this in an another commit ?

Comment (btmills) [inline docs/rules/no-magic-number.md]: If you make your change, then do `git commit -a --amend --no-edit`, that'll add the change to the current commit. You can then `git push -f` to force push to your `new-rule-no-magic-number` branch, which will automatically update this PR.

File: lib/rules/no-magic-number.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    return {
+        "Literal": function(node) {
+            var parent = node.parent,
+                okTypes = detectObjects ? [] : ["ObjectExpression", "Property", "AssignmentExpression"];
+
+            if (!isNumber(node)) {
+                return;
+            } else if (shouldIgnoreNumber(node.value)) {
+                return;
+            }
```
Comment (btmills) [inline lib/rules/no-magic-number.js]: You can combine this into one condition: `if (!isNumber(node) || shouldIgnoreNumber(node.value)) {`. That's how I'd do it personally; feel free to ignore.

File: lib/rules/no-magic-number.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
+/**
+ * @fileoverview Rule to flag statements that use magic numbers (adapted from https://github.com/danielstjules/buddy.js)
+ * @author Vincent Lemeunier
```
Comment (btmills) [inline lib/rules/no-magic-number.js]: Don't forget to add your copyright here and in the tests. ([example](http://eslint.org/docs/developer-guide/working-with-rules))

File: tests/lib/rules/no-magic-number.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                "function invokeInTen(func) {" +
+                  "setTimeout(func, 10);" +
+                "}",
+            env: { es6: true },
+            errors: [
+                { message: "No magic number: 10"},
+                { message: "No magic number: 10"},
+                { message: "No magic number: 24"},
+                { message: "No magic number: 1000"},
+                { message: "No magic number: 10"}
```
Comment (btmills) [inline tests/lib/rules/no-magic-number.js]: Can you add line numbers to the error messages for this test? The numbers in the other tests are unique, but we want to make sure it's still warning on the correct numbers here if someone makes changes to the rule later. ([example](https://github.com/eslint/eslint/blob/master/tests/lib/rules/arrow-parens.js#L49))

File: lib/rules/no-magic-number.js (hunk 3 above)
Comment (kombucha) [inline lib/rules/no-magic-number.js]: Done !

File: tests/lib/rules/no-magic-number.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
+/**
+ * @fileoverview Tests for complexity rule.
+ * @author Vincent Lemeunier
```
Comment (btmills) [inline tests/lib/rules/no-magic-number.js]: One last thing: can you add the copyright line for this file too?

**Comment by nzakas** [comment]
You mentioned in the issue that you took code from buddy.js. if that's true, then you need to include the license in this file: https://github.com/danielstjules/buddy.js/blob/master/LICENSE

**Comment by kombucha** [comment]
Yes, I didn't straight out copy/paste the code, but I did use it as a reference.
I'll put the license at the top of the main file.

**Comment by kombucha** [comment]
Hum seems like it causes an issue with the license validation ?

**Comment by btmills** [comment]
Not your fault. See #4031.

**Comment by sindresorhus** [comment]
Should be fixed now. Just rerun travis or push new commit.

**Comment by ilyavolodin** [comment]
LGTM

**Comment by nzakas** [comment]
Oops,  just noticed: should this be no-magic-numbers (plural)? Most of our rules seen to use plurals.

**Comment by ilyavolodin** [comment]
I think it should. Didn't notice that either.

**Comment by nzakas** [comment]
Okay, opened: https://github.com/eslint/eslint/issues/4053
